### DISCUSSION TITLE
Add support for displaying images/stickers in the terminal

### DISCUSSION THREAD

**Discussion by WeetHet** [d99kris/nchat#99] [Ideas]
Some terminal emulators support viewing images using imgcat (or other similar tools). Would be great to see support for those here (as a configuration option).

**Comment by d99kris** [comment]
Hi @WeetHet - thanks for suggesting a new feature. I'm a little hesitent to adding it, mainly because I personally prefer simple/plain TUI interfaces, and also because I suspect there could be some issues getting an image viewer library to play well with the ncurses framework used by nchat. One option might be to look into other clients like [telega.el](https://github.com/zevlg/telega.el) for emacs, not sure if it displays images inline in terminal mode though.

**Comment by WeetHet** [comment]
We still are supposed to view images somehow, and with this feature we could probably make it work without external image viewer

**Comment by itxasos23** [comment]
+1 to this feature, being able to quickly see images without opening them would be very helpful.
[felix](https://github.com/kyoheiu/felix) uses [chafa](https://github.com/hpjansson/chafa) for that, which is expected to be installed separately.

**Reply by d99kris** [reply]
Thanks for sharing! Looks like chafa provides some [sample code](https://github.com/hpjansson/chafa/blob/master/tests/ncurses.c) for getting it to work with ncurses, so I suppose it would be feasible. Might look into it, but cannot promise anything.

**Comment by d99kris** [comment]
This was also requested in #149 (which suggested to use ueberzugpp for implementation).

**Comment by Disonantemus** [comment]
I do like the [nnn](https://github.com/jarun/nnn) way with both simple shell plugins that open external programs and then return:
- [preview-tui](https://github.com/jarun/nnn/blob/master/plugins/preview-tui)
- [nuke](https://github.com/jarun/nnn/blob/master/plugins/nuke)

With many ways to open files, using features from: tmux, kitty, wezterm, etc ... to have options with GUI, TUI, TTY (viu), Framebuffer (fim, fbida). And not just xdg-open.
Or just show sticker/gif with a shell script or package and then return to nchat.

**Comment by Disonantemus** [comment]
Maybe something like (proposal/feature):

`Select Message > Alt-v > show reaction, open` with:
1. Default viewer on desktop (X11/xdg-open|wayland), or can be custom in config.
2. Show big 32x32 ANSI/ASCII art pop up, showing emoji. Can be done with: aalib, libcaca, viu, chafa, timg, etc.

[Similar with an Image](https://github.com/d99kris/nchat/discussions/153#discussioncomment-8053865)

**Comment by RayBytes** [comment]
just here to give a +1, would be a really cool feature to have

**Comment by DaemonLife** [comment]
It would be great to make this an optional feature.

**Comment by depocoder** [comment]
@d99kris hello big thanks for this project. Would you accept PR for this feature?

**Reply by d99kris** [reply]
Hello @depocoder - I think as long as it doesn't introduce too much complexity, and can be an optional feature enabled in config, I would consider a PR. Ideally code changes should mainly affect `src/uihistoryview.[cpp|h]` plus all dependency handling. From UI perspective I suspect it's better to always show inline images (when the feature is enabled) rather than requiring user action, and probably at a fixed height(?). From dependency perspective both Linux and Mac should be supported by the lib(s) picked. We likely need a compile-time feature-flag also (see `HAS_MULTIPROTOCOL` for an example) and we might need to have it disabled initially. I'm happy to discuss more if you have questions or ideas.

**Comment by fr0stb1rd** [comment]
Would be great

**Comment by desgua** [comment]
I want to share my workaround. I didn't make a pull request for this because (1) this solution works for Linux users on Wayland (not all users) and (2) it requires the user to know what they're doing. It isn't just a 'copy and paste' solution. It works great here and it may be useful as an idea for other users.
`chafa` developer has stated that chafa doesn't play well with ncurses. I share that opinion because it is annoying when it mess with the text and userinterface of nchat, even if it happens only 10-20% of the time.
This workaround uses swayimg instead.

First I configured **dwl** (my window manager with the customfloat patch) to always open `swayimg --appid=img_whats_view` at a specific position of the screen, matching exactly the top right of nchat. Here is the relevant part of my dwl config.h:
`{ "img_whats_view",  NULL,              0,         	1,           -1, 	  1300, 100, 600, 600 },`

then I made a bash script named `nchat_display` like this:
```
#!/usr/bin/dash

pidfile="/tmp/nchat_display.pid"

# Check if PID file exists and a process is running with the stored PID
if [ -f "$pidfile" ]; then
	pid="$(cat "$pidfile")"
	if $(kill -0 "$pid" >/dev/null 2>&1) ; then
		echo "Another instance of nchat_display is already running. Exiting..."
		exit 0
	fi
fi

if [ "$1" = 'close' ]; then
	# Close any the previously opened picture
	prev_pid="$(ps -ef | grep 'img_whats_view' | grep -v 'grep' | awk '{ print $2 }')"
	
	if [ "$prev_pid" ]; then 
		for pid in $prev_pid ; do
			kill -15 "$pid"
		done
	fi
	exit 0
fi

if [ ! -f "$1" ]; then
	# Silently exit if we don't find the file
	exit 0
fi

cache="/tmp/nchat_display_cache"
mkdir -p "$cache"
file_extension="$(printf '%s' "$1" | grep -oE "\.[a-zA-Z0-9]{3,4}$")"

# pdf
if [ "$file_extension" = '.pdf' ]; then 
	file_name="$(printf '%s' "$1" | awk -F '/' '{ print $NF }')"
	name="$cache/${file_name}-01.ppm"
	if [ ! -f "$name" ]; then
		name="$cache/${file_name}-1.ppm"
	fi

	if [ ! -f "$name" ]; then
		timeout .2 pdftoppm -f 1 -l 1 -r 150 "$1" "$cache/$file_name"
	fi

# video
elif [ "$(printf '%s' "$file_extension" | grep -E 'f4v|mpeg|mp4|webv')" ]; then 
	file_name="$(printf '%s' "$1" | awk -F '/' '{ print $NF }')"
	name="$cache/$(printf '%s' "$file_name" | sed 's/\.....$/\.jpg/ ; s/\....$/\.jpg/')"

	if [ ! -f "$name" ]; then
		timeout .2 ffmpeg -i "$1" -vframes 1 -q:v 1 "$name" >/dev/null 2>&1
	fi

# img
elif [ "$(printf '%s' "$file_extension" | grep -E 'jpg|jpeg|png|webp')" ]; then 
	name="$1"

# audio
elif [ "$(printf '%s' "$file_extension" | grep -E 'oga')" ]; then 
	# Silently exit
	exit 0

else
	notify "Extension not handled: $file_extension"
	exit 0
fi


# Open the image
swayimg --appid=img_whats_view --size=600x600 "$name" >/dev/null 2>&1 &


# Get back to nchat
sleep .3
wlrctl keyboard type k modifiers super

[ -f "$pidfile" ] && rm "$pidfile"
exit 0
```

And I modified nchat/src/uihistoryview.cpp like this:

```
diff --git a/src/uihistoryview.cpp b/src/uihistoryview.cpp
index ef9da404..1965fc76 100644
--- a/src/uihistoryview.cpp
+++ b/src/uihistoryview.cpp
@@ -85,6 +85,19 @@ void UiHistoryView::Draw()
   werase(m_PaddedWin);
   wbkgd(m_PaddedWin, attributeTextNormal | colorPairTextRecv | ' ');
 
+  // by desgua to display image
+  static std::string m_LastSelectedMsgId;
+  if (!m_Model->GetSelectMessageActiveLocked())
+  {
+    if (!m_LastSelectedMsgId.empty())
+    {
+      m_LastSelectedMsgId = "";
+      std::string command = "nchat_display close &";
+      system(command.c_str());
+    }
+  }
+  // end by desgua
+
   m_HistoryShowCount = 0;
 
   bool firstMessage = true;
@@ -132,6 +145,16 @@ void UiHistoryView::Draw()
       }
 
       wlines = StrUtil::WordWrap(StrUtil::ToWString(text), m_PaddedW, false, false, false, 2);
+
+      // by desgua to display image
+      if (isSelectedMessage && m_LastSelectedMsgId != msg.id)
+      {
+        m_LastSelectedMsgId = msg.id;
+          std::string command = "nchat_display close &";
+          system(command.c_str());
+      }
+      // end by desgua to display image
+
     }
 
     // Quoted message
@@ -222,6 +245,19 @@ void UiHistoryView::Draw()
 
       std::wstring fileStr = attachmentIndicator + StrUtil::ToWString(fileName + fileStatus);
       wlines.insert(wlines.begin(), fileStr);
+
+      // by desgua to display image
+      if (isSelectedMessage && m_LastSelectedMsgId != msg.id)
+      {
+        m_LastSelectedMsgId = msg.id;
+
+        if (fileInfo.fileStatus == FileStatusDownloaded)
+        {
+          std::string command = "nchat_display \"" + fileInfo.filePath + "\" &";
+          system(command.c_str());
+        }
+      }
+      // end by desgua to display image
     }
 
     // Reactions

```

**Reply by desgua** [reply]
ps. The image is displayed when the message with an attachment is selected. By the default, the shortcut to select a message is KeyUp.
